README.md
# LockAndCache

[![Build Status](https://travis-ci.org/seamusabshere/lock_and_cache.svg?branch=master&v=2.2.0)](https://travis-ci.org/seamusabshere/lock_and_cache)
[![Code Climate](https://codeclimate.com/github/seamusabshere/lock_and_cache/badges/gpa.svg?v=2.2.0)](https://codeclimate.com/github/seamusabshere/lock_and_cache)
[![Gem Version](https://badge.fury.io/rb/lock_and_cache.svg?v=2.2.0)](http://badge.fury.io/rb/lock_and_cache)
[![Security](https://hakiri.io/github/seamusabshere/lock_and_cache/master.svg?v=2.2.0)](https://hakiri.io/github/seamusabshere/lock_and_cache/master)
[![Inline docs](http://inch-ci.org/github/seamusabshere/lock_and_cache.svg?branch=master&v=2.2.0)](http://inch-ci.org/github/seamusabshere/lock_and_cache)

Lock and cache using redis!

Most caching libraries don't do locking, meaning that >1 process can be calculating a cached value at the same time. Since you presumably cache things because they cost CPU, database reads, or money, doesn't it make sense to lock while caching?

## Quickstart

```ruby
LockAndCache.lock_storage = Redis.new db: 3
LockAndCache.cache_storage = Redis.new db: 4

LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10, nil_expires: 1) do
  # get yer stock quote
  # if 50 processes call this at the same time, only 1 will call the stock quote service
  # the other 49 will wait on the lock, then get the cached value
  # the value will expire in 10 seconds
  # but if the value you get back is nil, that will expire after 1 second
end
```

## Sponsor

<p><a href="https://www.faraday.io"><img src="https://s3.amazonaws.com/faraday-assets/files/img/logo.svg" alt="Faraday logo"/></a></p>

We use [`lock_and_cache`](https://github.com/seamusabshere/lock_and_cache) for [B2C customer intelligence at Faraday](https://www.faraday.io).

## TOC

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


  - [Theory](#theory)
  - [Practice](#practice)
    - [Setup](#setup)
    - [Locking](#locking)
    - [Caching](#caching)
      - [Standalone mode](#standalone-mode)
      - [Context mode](#context-mode)
  - [Special features](#special-features)
    - [Locking of course!](#locking-of-course)
    - [Heartbeat](#heartbeat)
    - [Context mode](#context-mode-1)
    - [nil_expires](#nil_expires)
  - [Tunables](#tunables)
  - [Few dependencies](#few-dependencies)
  - [Wishlist](#wishlist)
  - [Contributing](#contributing)
- [Copyright](#copyright)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Theory

`lock_and_cache`...

1. <span style="color: red;">returns cached value</span> (if exists)
2. <span style="color: green;">acquires a lock</span>
3. <span style="color: red;">returns cached value</span> (just in case it was calculated while we were waiting for a lock)
4. <span style="color: red;">calculates and caches the value</span>
5. <span style="color: green;">releases the lock</span>
6. <span style="color: red;">returns the value</span>

As you can see, most caching libraries only take care of (1) and (4) (well, and (5) of course).

If an error is raised during calculation, that error is propagated to all waiters for 1 second.

## Practice

### Setup

```ruby
LockAndCache.lock_storage = Redis.new db: 3
LockAndCache.cache_storage = Redis.new db: 4
```

It will use this redis for both locking and storing cached values.

### Locking

Just uses Redis naive locking with NX.

A 32-second heartbeat is used that will clear the lock if a process is killed.

### Caching

This gem is a simplified, improved version of https://github.com/seamusabshere/cache_method. In that library, you could only cache a method call.

In this library, you have two options: providing the whole cache key every time (standalone) or letting the library pull information about its context.

```ruby
# standalone example
LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10) do
  # ...
end

# context example
def stock_price(date)
  lock_and_cache(date, expires: 10) do
    # ...
  end
end
def lock_and_cache_key
  company
end
```

#### Standalone mode

```ruby
LockAndCache.lock_and_cache(:stock_price, company: 'MSFT', date: '2015-05-05') do
  # get yer stock quote
end
```

You probably want an expiry

```ruby
LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10) do
  # get yer stock quote
end
```

Note how we separated options (`{expires: 10}`) from a hash that is part of the cache key (`{company: 'MSFT', date: '2015-05-05'}`).

One other crazy thing: `nil_expires` - for when you want to check more often if the external stock price service returned nil

```ruby
LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10, nil_expires: 1) do
  # get yer stock quote
end
```

Clear it with

```ruby
LockAndCache.clear :stock_price, company: 'MSFT', date: '2015-05-05'
```

Check locks with

```ruby
LockAndCache.locked? :stock_price, company: 'MSFT', date: '2015-05-05'
```

Caching, locking or both can be bypassed

```ruby
LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10, bypass: :cache) do
  # ignore any cached value, get yer stock quote
end
```

```ruby
LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10, bypass: :lock) do
  # return cached value if it exists, otherwise get yer stock quote *without* acquiring lock
end
```

```ruby
LockAndCache.lock_and_cache(:stock_price, {company: 'MSFT', date: '2015-05-05'}, expires: 10, bypass: :both) do
  # get yer stock quote without caching or locking
end
```

#### Context mode

"Context mode" simply adds the class name, method name, and context key (the results of `#id` or `#lock_and_cache_key`) of the caller to the cache key.

```ruby
class Stock
  include LockAndCache

  def initialize(company)
    [...]
  end

  def stock_price(date)
    lock_and_cache(date, expires: 10) do
      # the cache key will be StockQuote (the class) + get (the method name) + id (the instance identifier) + date (the arg you specified)
    end
  end

  def lock_and_cache_key # <---------- if you don't define this, it will try to call #id
    company
  end
end
```

The cache key will be StockQuote (the class) + get (the method name) + id (the instance identifier) + date (the arg you specified).

In other words, it auto-detects the class, method, context key ... and you add other args if you want.

Clear it with

```ruby
blog.lock_and_cache_clear(:get, date)
```

## Special features

### Locking of course!

Most caching libraries don't do locking, meaning that >1 process can be calculating a cached value at the same time. Since you presumably cache things because they cost CPU, database reads, or money, doesn't it make sense to lock while caching?

### Heartbeat

If the process holding the lock dies, we automatically remove the lock so somebody else can do it (using heartbeats).

### Context mode

This pulls information about the context of a lock_and_cache block from the surrounding class, method, and object... so that you don't have to!

Standalone mode is cool too, tho.

### nil_expires

You can expire nil values with a different timeout (`nil_expires`) than other values (`expires`).

## Tunables

* `LockAndCache.lock_storage=[redis]`
* `LockAndCache.cache_storage=[redis]`
* `ENV['LOCK_AND_CACHE_DEBUG']='true'` if you want some debugging output on `$stderr`

## Few dependencies

* [activesupport](https://rubygems.org/gems/activesupport) (come on, it's the bomb)
* [redis](https://github.com/redis/redis-rb)

## Known issues

* In cache keys, can't distinguish {a: 1} from [[:a, 1]]

## Wishlist

* Convert most tests to use standalone mode, which is easier to understand
* Check options
* Lengthen heartbeat so it's not so sensitive
* Clarify which options are seconds or milliseconds

## Contributing

1. Fork it ( https://github.com/[my-github-username]/lock_and_cache/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

# Copyright

Copyright 2015 Seamus Abshere


CHANGELOG
6.0.1

* Enhancements

  * Use Redis#exists? to avoid deprec warning

6.0.0

* Breaking changes

  * Set lock_storage and cache_storage separately

5.0.0

* Enhancements / breaking changes

  * Propagate errors to all waiters up to 1 second after the error
  * Stop using redlock, just use plain single-node Redis locking

4.0.6

* ?

  * Don't test on ruby 2.1

* Enhancements

  * LockAndCache.cached?(*key_parts) to check if a value is cached

4.0.5 / 2017-04-01

* Enhancements

  * allow dates and times in keys
  * Test on ruby 2.3.0 and 2.4.1
  * 2x faster key generation

4.0.4 / 2016-04-11

* Bug fixes

  * Don't default to debug logging

4.0.3 / 2016-04-11

* Bug fixes

  * Allow true or false in keys

4.0.2 / 2016-04-11

* Bug fixes

  * When generating key, recurse into #lock_and_cache_key

4.0.1 / 2016-04-11

* Bug fixes

  * Avoid deadlocks related to logging

4.0.0 / 2016-04-11

* Breaking changes

  * The cache key generation I've always wanted: recursively call #id or #lock_and_cache_key

3.0.1 / 2016-04-04

* Enhancements

  * Don't use deprecated Thread.exclusive

3.0.0 / 2016-04-02

* Breaking changes

  * In context mode (when you `include LockAndCache`), really call #lock_and_cache_key or #id on the instance

2.2.2 / 2015-12-18

* Bug fixes

  * Don't die if you pass a non-integer expires - round it

2.2.1 / 2015-12-14

* Bug fixes

  * Construct key using the correct class name

2.2.0 / 2015-11-15

* Enhancements

  * Increase default heartbeat expires to 32 seconds from 2 (which was too strict IMO)
  * Allow setting heartbeat_expires: globally (LockAndCache.heartbeat_expires=) or per call
  * Provide LockAndCache.locked?()

2.1.1 / 2015-10-26

* Bug fixes

  * Blow up if you try to use standalone mode without a key

2.1.0 / 2015-10-26

* Enhancements

  * Better documentation
  * Standalone mode (LockAndCache.lock_and_cache([...]) {})
  * Nulls can be set to expire sooner than non-null return values (`nil_expires`)

2.0.2 / 2015-10-16

* Bug fixes (?)

  * Make sure cached values are valid marshal format (seen in the wild that they're nil)

* Enhancements

  * Use original redlock gem now that it supports extend

2.0.1 / 2015-09-14

* Bug fixes

  * Don't explicitly kill the lock extender thread because that sometimes causes deadlocks (don't know why)

2.0.0 / 2015-09-11

* Breaking changes

  * Stricter key digest - differentiates symbols and strings
  * No more lock_expires or lock_spin options

* Bug fixes

  * Allow method names with non-word chars like #foo?

* Enhancements

  * heartbeats so that SIGKILL will effectively clear the lock
  * #lock_and_cache_clear now clears lock too

1.1.0 / 2015-08-07

* Breaking changes

  * Reduce default lock expiry to 1 day instead of weird 3 days

* Enhancements

  * Added :max_lock_wait option inspired by @leandromoreira

1.0.3 / 2015-08-06

* Enhancements

  * More granular debug output

1.0.2 / 2015-08-06

* Bug fixes

  * Put LockAndCache.flush back

1.0.1 / 2015-08-06

* Bug fixes

  * Return value properly if lock was acquired but cached value immediately found

* Enhancements

  * Documentation

1.0.0 / 2015-08-05

* Enhancements

  * Use Redis redlock http://redis.io/topics/distlock instead of Postgres advisory locks
  * No more dependency on ActiveRecord or Postgres!

0.1.2 / 2015-06-24

* Enhancements

  * Add :expires option in seconds

0.1.1 / 2015-02-04

* Enhancements

  * Clear individual cached things with #lock_and_cache_clear

0.1.0 / 2015-01-22

* Breaking changes

  * Redis only
  * Now you use it inside methods (like Rails.cache.fetch) instead of outside (like cache_method)

* Enhancements

  * Way simpler, no dependency on CacheMethod

0.0.5 / 2014-12-12

* Enhancements

  * ENV['LOCK_AND_CACHE_DEBUG'] == 'true' debug output to $stderr

0.0.4 / 2014-12-12

* Bug fixes

  * Pass arguments while caching method results

0.0.3 / 2014-12-12

* Enhancements

  * Save a trip to the database if something is already cached

0.0.2 / 2014-12-11

* Bug fixes

  * Gem name is activerecord

0.0.1 / 2014-12-11

initial release!


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)
require 'lock_and_cache'

require 'timeout'

require 'redis'
LockAndCache.lock_storage = Redis.new db: 3
LockAndCache.cache_storage = Redis.new db: 4

require 'thread/pool'

require 'pry'


spec/lock_and_cache_spec.rb
require 'spec_helper'

class Foo
  include LockAndCache

  def initialize(id)
    @id = id
    @count = 0
    @count_exp = 0
    @click_single_hash_arg_as_options = 0
    @click_last_hash_as_options = 0
  end

  def click
    lock_and_cache do
      @count += 1
    end
  end

  def cached_rand
    lock_and_cache do
      rand
    end
  end

  def click_null
    lock_and_cache do
      nil
    end
  end

  def click_exp
    lock_and_cache(expires: 1) do
      @count_exp += 1
    end
  end

  # foo will be treated as option, so this is cacheable
  def click_single_hash_arg_as_options
    lock_and_cache(foo: rand, expires: 1) do
      @click_single_hash_arg_as_options += 1
    end
  end

  # foo will be treated as part of cache key, so this is uncacheable
  def click_last_hash_as_options
    lock_and_cache({foo: rand}, expires: 1) do
      @click_last_hash_as_options += 1
    end
  end

  def lock_and_cache_key
    @id
  end
end

class FooId
  include LockAndCache
  def click
    lock_and_cache do
      nil
    end
  end
  def id
    @id ||= rand
  end
end

class FooClass
  class << self
    include LockAndCache
    def click
      lock_and_cache do
        nil
      end
    end
    def id
      raise "called id"
    end
  end
end

require 'set'
$clicking = Set.new
class Bar
  include LockAndCache

  def initialize(id)
    @id = id
    @count = 0
    @mutex = Mutex.new
  end

  def unsafe_click
    @mutex.synchronize do
      # puts "clicking bar #{@id} - #{$clicking.to_a} - #{$clicking.include?(@id)} - #{@id == $clicking.to_a[0]}"
      raise "somebody already clicking Bar #{@id}" if $clicking.include?(@id)
      $clicking << @id
    end
    sleep 1
    @count += 1
    $clicking.delete @id
    @count
  end

  def click
    lock_and_cache do
      unsafe_click
    end
  end

  def slow_click
    lock_and_cache do
      sleep 1
    end
  end

  def lock_and_cache_key
    @id
  end
end

class Sleeper
  include LockAndCache

  def initialize
    @id = SecureRandom.hex
  end

  def poke
    lock_and_cache heartbeat_expires: 2 do
      sleep
    end
  end

  def lock_and_cache_key
    @id
  end
end

describe LockAndCache do
  before do
    LockAndCache.flush_locks
    LockAndCache.flush_cache
  end

  it 'has a version number' do
    expect(LockAndCache::VERSION).not_to be nil
  end

  describe "caching" do
    let(:foo) { Foo.new(rand.to_s) }
    it "works" do
      expect(foo.click).to eq(1)
      expect(foo.click).to eq(1)
    end

    it "can be cleared" do
      expect(foo.click).to eq(1)
      foo.lock_and_cache_clear :click
      expect(foo.click).to eq(2)
    end

    it "can be expired" do
      expect(foo.click_exp).to eq(1)
      expect(foo.click_exp).to eq(1)
      sleep 1.5
      expect(foo.click_exp).to eq(2)
    end

    it "can cache null" do
      expect(foo.click_null).to eq(nil)
      expect(foo.click_null).to eq(nil)
    end

    it "treats single hash arg as options" do
      expect(foo.click_single_hash_arg_as_options).to eq(1)
      expect(foo.click_single_hash_arg_as_options).to eq(1)
      sleep 1.1
      expect(foo.click_single_hash_arg_as_options).to eq(2)
    end

    it "treats last hash as options" do
      expect(foo.click_last_hash_as_options).to eq(1)
      expect(foo.click_last_hash_as_options).to eq(2) # it's uncacheable to prove we're not using as part of options
      expect(foo.click_last_hash_as_options).to eq(3)
    end

    it "calls #lock_and_cache_key" do
      expect(foo).to receive(:lock_and_cache_key)
      foo.click
    end

    it "calls #lock_and_cache_key to differentiate" do
      a = Foo.new 1
      b = Foo.new 2
      expect(a.cached_rand).not_to eq(b.cached_rand)
    end
  end

  describe 'self-identification in context mode' do
    it "calls #id for non-class" do
      foo_id = FooId.new
      expect(foo_id).to receive(:id)
      foo_id.click
    end
    it "calls class name for non-class" do
      foo_id = FooId.new
      expect(FooId).to receive(:name)
      foo_id.click
    end
    it "uses class name for class" do
      expect(FooClass).to receive(:name)
      expect(FooClass).not_to receive(:id)
      FooClass.click
    end
  end

  describe "locking" do
    let(:bar) { Bar.new(rand.to_s) }

    it "it blows up normally (simple thread)" do
      a = Thread.new do
        bar.unsafe_click
      end
      b = Thread.new do
        bar.unsafe_click
      end
      expect do
        a.join
        b.join
      end.to raise_error(/somebody/)
    end

    it "it blows up (pre-existing thread pool, more reliable)" do
      pool = Thread.pool 2
      Thread::Pool.abort_on_exception = true
      expect do
        pool.process do
          bar.unsafe_click
        end
        pool.process do
          bar.unsafe_click
        end
        pool.shutdown
      end.to raise_error(/somebody/)
    end

    it "doesn't blow up if you lock it (simple thread)" do
      a = Thread.new do
        bar.click
      end
      b = Thread.new do
        bar.click
      end
      a.join
      b.join
    end

    it "doesn't blow up if you lock it (pre-existing thread pool, more reliable)" do
      pool = Thread.pool 2
      Thread::Pool.abort_on_exception = true
      pool.process do
        bar.click
      end
      pool.process do
        bar.click
      end
      pool.shutdown
    end

    it "can set a wait time" do
      pool = Thread.pool 2
      Thread::Pool.abort_on_exception = true
      begin
        old_max = LockAndCache.max_lock_wait
        LockAndCache.max_lock_wait = 0.5
        expect do
          pool.process do
            bar.slow_click
          end
          pool.process do
            bar.slow_click
          end
          pool.shutdown
        end.to raise_error(LockAndCache::TimeoutWaitingForLock)
      ensure
        LockAndCache.max_lock_wait = old_max
      end
    end

    it 'unlocks if a process dies' do
      child = nil
      begin
        sleeper = Sleeper.new
        child = fork do
          sleeper.poke
        end
        sleep 0.1
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(true)  # the other process has it
        Process.kill 'KILL', child
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(true)  # the other (dead) process still has it
        sleep 2
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(false) # but now it should be cleared because no heartbeat
      ensure
        Process.kill('KILL', child) rescue Errno::ESRCH
      end
    end

    it "pays attention to heartbeats" do
      child = nil
      begin
        sleeper = Sleeper.new
        child = fork do
          sleeper.poke
        end
        sleep 0.1
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(true) # the other process has it
        sleep 2
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(true) # the other process still has it
        sleep 2
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(true) # the other process still has it
        sleep 2
        expect(sleeper.lock_and_cache_locked?(:poke)).to eq(true) # the other process still has it
      ensure
        Process.kill('TERM', child) rescue Errno::ESRCH
      end
    end

  end

  describe 'standalone' do
    it 'works like you expect' do
      count = 0
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(1)
      expect(count).to eq(1)
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(1)
      expect(count).to eq(1)
    end

    it 'really caches' do
      expect(LockAndCache.lock_and_cache('hello') { :red }).to eq(:red)
      expect(LockAndCache.lock_and_cache('hello') { raise(Exception.new("stop")) }).to eq(:red)
    end

    it 'doesn\'t break when bypass has an unknown value' do
      expect(LockAndCache.lock_and_cache('hello', bypass: nil) { :red }).to eq(:red)
      expect(LockAndCache.lock_and_cache('hello', bypass: :foo) { raise(Exception.new("stop")) }).to eq(:red)
    end

    it 'doesn\'t cache when bypass == :cache' do
      count = 0
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(1)
      expect(count).to eq(1)
      expect(LockAndCache.lock_and_cache('hello', bypass: :cache) { count += 1 }).to eq(2)
      expect(count).to eq(2)
    end

    it 'doesn\'t cache when bypass == :both' do
      count = 0
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(1)
      expect(count).to eq(1)
      expect(LockAndCache.lock_and_cache('hello', bypass: :both) { count += 1 }).to eq(2)
      expect(count).to eq(2)
    end

    it 'caches errors (briefly)' do
      count = 0
      expect {
        LockAndCache.lock_and_cache('hello') { count += 1; raise("stop") }
      }.to raise_error(/stop/)
      expect(count).to eq(1)
      expect {
        LockAndCache.lock_and_cache('hello') { count += 1; raise("no no not me") }
      }.to raise_error(/LockAndCache.*stop/)
      expect(count).to eq(1)
      sleep 1
      expect {
        LockAndCache.lock_and_cache('hello') { count += 1; raise("retrying") }
      }.to raise_error(/retrying/)
      expect(count).to eq(2)
    end

    it "can be queried for cached?" do
      expect(LockAndCache.cached?('hello')).to be_falsy
      LockAndCache.lock_and_cache('hello') { nil }
      expect(LockAndCache.cached?('hello')).to be_truthy
    end

    it 'allows expiry' do
      count = 0
      expect(LockAndCache.lock_and_cache('hello', expires: 1) { count += 1 }).to eq(1)
      expect(count).to eq(1)
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(1)
      expect(count).to eq(1)
      sleep 1.1
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(2)
      expect(count).to eq(2)
    end

    it "allows float expiry" do
      expect{LockAndCache.lock_and_cache('hello', expires: 1.5) {}}.not_to raise_error
    end

    it 'can be nested' do
      expect(LockAndCache.lock_and_cache('hello') do
        LockAndCache.lock_and_cache('world') do
          LockAndCache.lock_and_cache('privyet') do
            123
          end
        end
      end).to eq(123)
    end

    it "requires a key" do
      expect do
        LockAndCache.lock_and_cache do
          raise "this won't happen"
        end
      end.to raise_error(/need/)
    end

    it 'allows checking locks' do
      expect(LockAndCache.locked?(:sleeper)).to be_falsey
      t = Thread.new do
        LockAndCache.lock_and_cache(:sleeper) { sleep 1 }
      end
      sleep 0.2
      expect(LockAndCache.locked?(:sleeper)).to be_truthy
      t.join
    end

    it 'allows clearing' do
      count = 0
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(1)
      expect(count).to eq(1)
      LockAndCache.clear('hello')
      expect(LockAndCache.lock_and_cache('hello') { count += 1 }).to eq(2)
      expect(count).to eq(2)
    end

    it 'allows clearing (complex keys)' do
      count = 0
      expect(LockAndCache.lock_and_cache('hello', {world: 1}, expires: 100) { count += 1 }).to eq(1)
      expect(count).to eq(1)
      LockAndCache.clear('hello', world: 1)
      expect(LockAndCache.lock_and_cache('hello', {world: 1}, expires: 100) { count += 1 }).to eq(2)
      expect(count).to eq(2)
    end

    it 'allows multi-part keys' do
      count = 0
      expect(LockAndCache.lock_and_cache(['hello', 1, { target: 'world' }]) { count += 1 }).to eq(1)
      expect(count).to eq(1)
      expect(LockAndCache.lock_and_cache(['hello', 1, { target: 'world' }]) { count += 1 }).to eq(1)
      expect(count).to eq(1)
    end

    it 'treats a single hash arg as a cache key (not as options)' do
      count = 0
      LockAndCache.lock_and_cache(hello: 'world', expires: 100) { count += 1 }
      expect(count).to eq(1)
      LockAndCache.lock_and_cache(hello: 'world', expires: 100) { count += 1 }
      expect(count).to eq(1)
      LockAndCache.lock_and_cache(hello: 'world', expires: 200) { count += 1 } # expires is being treated as part of cache key
      expect(count).to eq(2)
    end

    it "correctly identifies options hash" do
      count = 0
      LockAndCache.lock_and_cache({ hello: 'world' }, expires: 1, ignored: rand) { count += 1 }
      expect(count).to eq(1)
      LockAndCache.lock_and_cache({ hello: 'world' }, expires: 1, ignored: rand) { count += 1 } # expires is not being treated as part of cache key
      expect(count).to eq(1)
      sleep 1.1
      LockAndCache.lock_and_cache({ hello: 'world' }) { count += 1 }
      expect(count).to eq(2)
    end
  end

  describe "shorter expiry for null results" do
    it "optionally caches null for less time" do
      count = 0
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; nil }
      expect(count).to eq(1)
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; nil }
      expect(count).to eq(1)
      sleep 1.1 # this is enough to expire
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; nil }
      expect(count).to eq(2)
    end

    it "normally caches null for the same amount of time" do
      count = 0
      expect(LockAndCache.lock_and_cache('hello', expires: 1) { count += 1; nil }).to be_nil
      expect(count).to eq(1)
      expect(LockAndCache.lock_and_cache('hello', expires: 1) { count += 1; nil }).to be_nil
      expect(count).to eq(1)
      sleep 1.1
      expect(LockAndCache.lock_and_cache('hello', expires: 1) { count += 1; nil }).to be_nil
      expect(count).to eq(2)
    end

    it "caches non-null for normal time" do
      count = 0
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; true }
      expect(count).to eq(1)
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; true }
      expect(count).to eq(1)
      sleep 1.1
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; true }
      expect(count).to eq(1)
      sleep 1
      LockAndCache.lock_and_cache('hello', nil_expires: 1, expires: 2) { count += 1; true }
      expect(count).to eq(2)
    end
  end


end


lib/lock_and_cache.rb
require 'logger'
require 'timeout'
require 'digest/sha1'
require 'base64'
require 'redis'
require 'active_support'
require 'active_support/core_ext'

require_relative 'lock_and_cache/version'
require_relative 'lock_and_cache/action'
require_relative 'lock_and_cache/key'

# Lock and cache using redis!
#
# Most caching libraries don't do locking, meaning that >1 process can be calculating a cached value at the same time. Since you presumably cache things because they cost CPU, database reads, or money, doesn't it make sense to lock while caching?
module LockAndCache
  DEFAULT_MAX_LOCK_WAIT = 60 * 60 * 24 # 1 day in seconds

  DEFAULT_HEARTBEAT_EXPIRES = 32 # 32 seconds

  class TimeoutWaitingForLock < StandardError; end

  # @param redis_connection [Redis] A redis connection to be used for lock storage
  def LockAndCache.lock_storage=(redis_connection)
    raise "only redis for now" unless redis_connection.class.to_s == 'Redis'
    @lock_storage = redis_connection
  end

  # @return [Redis] The redis connection used for lock and cached value storage
  def LockAndCache.lock_storage
    @lock_storage
  end

  # @param redis_connection [Redis] A redis connection to be used for cached value storage
  def LockAndCache.cache_storage=(redis_connection)
    raise "only redis for now" unless redis_connection.class.to_s == 'Redis'
    @cache_storage = redis_connection
  end

  # @return [Redis] The redis connection used for cached value storage
  def LockAndCache.cache_storage
    @cache_storage
  end

  # @param logger [Logger] A logger.
  def LockAndCache.logger=(logger)
    @logger = logger
  end

  # @return [Logger] The logger.
  def LockAndCache.logger
    @logger
  end

  # Flush LockAndCache's cached value storage.
  #
  # @note If you are sharing a redis database, it will clear it...
  #
  # @note If you want to clear a single key, try `LockAndCache.clear(key)` (standalone mode) or `#lock_and_cache_clear(method_id, *key_parts)` in context mode.
  def LockAndCache.flush_cache
    cache_storage.flushdb
  end

  # Flush LockAndCache's lock storage.
  #
  # @note If you are sharing a redis database, it will clear it...
  #
  # @note If you want to clear a single key, try `LockAndCache.clear(key)` (standalone mode) or `#lock_and_cache_clear(method_id, *key_parts)` in context mode.
  def LockAndCache.flush_locks
    lock_storage.flushdb
  end

  # Lock and cache based on a key.
  #
  # @param key_parts [*] Parts that should be used to construct a key.
  #
  # @note Standalone mode. See also "context mode," where you mix LockAndCache into a class and call it from within its methods.
  #
  # @note A single hash arg is treated as a cache key, e.g. `LockAndCache.lock_and_cache(foo: :bar, expires: 100)` will be treated as a cache key of `foo: :bar, expires: 100` (which is probably wrong!!!). Try `LockAndCache.lock_and_cache({ foo: :bar }, expires: 100)` instead. This is the opposite of context mode.
  def LockAndCache.lock_and_cache(*key_parts_and_options, &blk)
    options = (key_parts_and_options.last.is_a?(Hash) && key_parts_and_options.length > 1) ? key_parts_and_options.pop : {}
    raise "need a cache key" unless key_parts_and_options.length > 0
    key = LockAndCache::Key.new key_parts_and_options
    action = LockAndCache::Action.new key, options, blk
    action.perform
  end

  # Clear a single key
  #
  # @note Standalone mode. See also "context mode," where you mix LockAndCache into a class and call it from within its methods.
  def LockAndCache.clear(*key_parts)
    key = LockAndCache::Key.new key_parts
    key.clear
  end

  # Check if a key is locked
  #
  # @note Standalone mode. See also "context mode," where you mix LockAndCache into a class and call it from within its methods.
  def LockAndCache.locked?(*key_parts)
    key = LockAndCache::Key.new key_parts
    key.locked?
  end

  # Check if a key is cached already
  #
  # @note Standalone mode. See also "context mode," where you mix LockAndCache into a class and call it from within its methods.
  def LockAndCache.cached?(*key_parts)
    key = LockAndCache::Key.new key_parts
    key.cached?
  end

  # @param seconds [Numeric] Maximum wait time to get a lock
  #
  # @note Can be overridden by putting `max_lock_wait:` in your call to `#lock_and_cache`
  def LockAndCache.max_lock_wait=(seconds)
    @max_lock_wait = seconds.to_f
  end

  # @private
  def LockAndCache.max_lock_wait
    @max_lock_wait || DEFAULT_MAX_LOCK_WAIT
  end

  # @param seconds [Numeric] How often a process has to heartbeat in order to keep a lock
  #
  # @note Can be overridden by putting `heartbeat_expires:` in your call to `#lock_and_cache`
  def LockAndCache.heartbeat_expires=(seconds)
    memo = seconds.to_f
    raise "heartbeat_expires must be greater than 2 seconds" unless memo >= 2
    @heartbeat_expires = memo
  end

  # @private
  def LockAndCache.heartbeat_expires
    @heartbeat_expires || DEFAULT_HEARTBEAT_EXPIRES
  end

  # Check if a method is locked on an object.
  #
  # @note Subject mode - this is expected to be called on an object whose class has LockAndCache mixed in. See also standalone mode.
  def lock_and_cache_locked?(method_id, *key_parts)
    key = LockAndCache::Key.new key_parts, context: self, method_id: method_id
    key.locked?
  end

  # Clear a lock and cache given exactly the method and exactly the same arguments
  #
  # @note Subject mode - this is expected to be called on an object whose class has LockAndCache mixed in. See also standalone mode.
  def lock_and_cache_clear(method_id, *key_parts)
    key = LockAndCache::Key.new key_parts, context: self, method_id: method_id
    key.clear
  end

  # Lock and cache a method given key parts.
  #
  # The cache key will automatically include the class name of the object calling it (the context!) and the name of the method it is called from.
  #
  # @param key_parts_and_options [*] Parts that you want to include in the lock and cache key. If the last element is a Hash, it will be treated as options.
  #
  # @return The cached value (possibly newly calculated).
  #
  # @note Subject mode - this is expected to be called on an object whose class has LockAndCache mixed in. See also standalone mode.
  #
  # @note A single hash arg is treated as an options hash, e.g. `lock_and_cache(expires: 100)` will be treated as options `expires: 100`. This is the opposite of standalone mode.
  def lock_and_cache(*key_parts_and_options, &blk)
    options = key_parts_and_options.last.is_a?(Hash) ? key_parts_and_options.pop : {}
    key = LockAndCache::Key.new key_parts_and_options, context: self, caller: caller
    action = LockAndCache::Action.new key, options, blk
    action.perform
  end
end

logger = Logger.new $stderr
logger.level = (ENV['LOCK_AND_CACHE_DEBUG'] == 'true') ? Logger::DEBUG : Logger::INFO
LockAndCache.logger = logger
